File: commons-core/src/test/java/xyz/wagyourtail/commons/util/TestCollectionUtils.java
package xyz.wagyourtail.commons.util;

import org.junit.jupiter.api.Test;
import xyz.wagyourtail.commons.core.CollectionUtils;

import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

public class TestCollectionUtils {

    public void testResolveArgs() {
        var map = Map.of(
                "input", "example.jar",
                "output", "output.jar",
                "wasd", "qwer"
        );
        var args = List.of(
                "{input}",
                "--output={output}",
                "{wasd}",
                "nothing"
        );

        var result = CollectionUtils.resolveArgs(map, args, false);

        assertEquals(
                List.of(
                        "example.jar",
                        "--output=output.jar",
                        "qwer",
                        "nothing"
                ),
                result
        );
    }

    public void testResolveArgs2() {
        var map = Map.of(
                "input", "example.jar",
                "output", "output.jar",
                "wasd", "qwer"
        );
        var args = List.of(
                "${input}",
                "--output=${output}",
                "${wasd}",
                "nothing"
        );

        var result = CollectionUtils.resolveArgs(map, args, true);

        assertEquals(
                List.of(
                        "example.jar",
                        "--output=output.jar",
                        "qwer",
                        "nothing"
                ),
                result
        );
    }

    @Test
    public void testMissingArg() {
        assertThrows(IllegalArgumentException.class, () -> {
            var map = Map.of(
                    "input", "example.jar",
                    "output", "output.jar"
            );
            var args = List.of(
                    "${input}",
                    "--output=${output}",
                    "${wasd}",
                    "nothing"
            );

            CollectionUtils.resolveArgs(map, args, true);
        }, "Property wasd not found");
    }

}

File: commons-kt-kt/src/commonTest/kotlin/xyz/wagyourtail/commonskt/test/utils/TestCollectionUtils.kt
package xyz.wagyourtail.commonskt.test.utils

import xyz.wagyourtail.commonskt.utils.resolveArgs
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class TestCollectionUtils {

    @Test
    fun testResolveArgs() {
        val map = mapOf(
            "input" to "example.jar",
            "output" to "output.jar",
            "wasd" to "qwer"
        )

        val args = listOf(
            "{input}",
            "--output={output}",
            "{wasd}",
            "nothing"
        )

        val expected = listOf(
            "example.jar",
            "--output=output.jar",
            "qwer",
            "nothing"
        )

        assertEquals(expected, map.resolveArgs(args, false))
    }

    @Test
    fun testResolveArgsWithDollar() {
        val map = mapOf(
            "input" to "example.jar",
            "output" to "output.jar",
            "wasd" to "qwer"
        )

        val args = listOf(
            "\${input}",
            "--output=\${output}",
            "\${wasd}",
            "nothing"
        )

        val expected = listOf(
            "example.jar",
            "--output=output.jar",
            "qwer",
            "nothing"
        )

        assertEquals(expected, map.resolveArgs(args, true))
    }

    @Test
    fun testMissingArg() {
        val map = mapOf(
            "input" to "example.jar",
            "output" to "output.jar",
        )

        val args = listOf(
            "\${input}",
            "--output=\${output}",
            "\${wasd}",
            "nothing"
        )

        assertFailsWith<IllegalArgumentException>("Property wasd not found") {
            map.resolveArgs(args, true)
        }
    }

}